test/events.js

```diff
@@ -0,0 +1,123 @@
+const fs = require('fs');
+const path = require('path');
+
+const db = require('../lib/db');
+const events = require('../lib/events');
+const P = require('../lib/promise');
+
+const UID = crypto.randomBytes(16).toString('hex');
+const mock = require('./lib/mock')({ userid: UID });
+const Server = require('./lib/server');
+const Static = require('./lib/static');
+
+const imagePath = path.join(__dirname, 'lib', 'firefox.png');
+const imageData = fs.readFileSync(imagePath);
+
+const SIZES = require('../lib/img').SIZES;
+const SIZE_SUFFIXES = Object.keys(SIZES).map(function(val) {
+  if (val === 'default') {
+    return '';
+  }
+  return '_' + val;
+});
+
+
+/*global describe,it,beforeEach,afterEach*/
+
+afterEach(function() {
+  mock.done();
+});
+
+
+describe('events', function() {
+  describe('onDeleteMessage', function() {
+    var tok = crypto.randomBytes(32).toString('hex');
+    function Message(type, onDel) {
+      if (typeof type === 'function') {
+        onDel = type;
+        type = 'delete';
+      }
+      return {
+        event: type,
+        uid: UID + '@accounts.firefox.com',
+        del: onDel
+      };
+    }
+
+    describe('avatar', function() {
+      beforeEach(function() {
+        mock.token({
+          user: UID,
+          email: 'user@example.domain',
+          scope: ['profile:avatar:write']
+        });
+        mock.image();
+        return Server.api.post({
+          url: '/avatar/upload',
+          payload: imageData,
+          headers: { authorization: 'Bearer ' + tok,
+            'content-type': 'image/png',
+            'content-length': imageData.length
+          }
+        }).then(function(res) {
+          assert.equal(res.statusCode, 201);
+          assert(res.result.url);
+          assert(res.result.id);
+          return res.result.url;
+        }).then(function(s3url) {
+          return P.all(SIZE_SUFFIXES).map(function(suffix) {
+            return Static.get(s3url + suffix);
+          });
+        }).then(function(responses) {
+          assert.equal(responses.length, SIZE_SUFFIXES.length);
+          responses.forEach(function(res) {
+            assert.equal(res.statusCode, 200);
+          });
+          mock.done();
+        });
+      });
+
+      it('should delete avatars', function(done) {
+        mock.deleteImage();
+        events.onData(new Message(function() {
+          db.getAvatars(UID).then(function(avatars) {
+            assert.equal(avatars.length, 0);
+          }).done(done, done);
+        }));
+      });
+
+      it('should not delete message on error', function(done) {
+        mock.workerFailure('delete');
+        events.onData(new Message(function() {
+          assert(false, 'message.del() should not be called');
+        }));
+        mock.log('events', function(record) {
+          return record.levelname === 'ERROR'
+              && record.args[0] === 'removeUser';
+        });
+
+        setTimeout(function() {
+          done();
+        }, 50);
```

Comment:
`onData` is set as the event handler, and nodejs inspects the return value of those. Also, with Promises, you can either return them, saying "it's someone else's responsibility now", or you can call `p.done()`, which will throw an error if it hasn't been previously handled. I'd rather call `done` inside that event handler, than hoping somewhere else does.

However, I can move the timeout into the log mock, since that's a requirement for the test to pass, and happens at the end.